### PR TITLE
feat: Add kustomize edit set replicas (#11650)

### LINKED ISSUE #11650: Add a replicas configuration option for kustomize
# Summary

Currently, kustomize configuration options are available only for `namePrefix`, `nameSuffix`, `images`, `commonLabels`, `commonAnnotations` according to [argocd-doc](https://argo-cd.readthedocs.io/en/stable/user-guide/kustomize/).

I would like to see a replicas transformer to change a number of replicas [kustomize-doc](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/replicas/).

# Motivation

I would like to change the number of replicas in `application.yaml`. Like,

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
spec:
    # kustomize specific config
    kustomize:
      replicas:
        - name: deployment-name
          count: 5
```

# Proposal

I think it is straightforward.

* Kustomize already implements `kustomize edit set replicas deployment-name=5`.
* And our code is ready to add the above command in https://github.com/argoproj/argo-cd/blob/b4c12815f0131b337404e67534eca54004ba47de/util/kustomize/kustomize.go#L87
* So, we can implement this feature by following the same way as #4613

### PR DESCRIPTION
**Pull request by toyamagu-2021**
Closes #11650

## Description

* Add code for kustomize transformer `kustomize edit set replicas my-deployment=2`
* I'd like to focus on CLI in this PR, so I don't make a change on UI.

## Test

Although I have added e2e  tests, we can test manually:

* By command: `dist/argocd app create kustomize --project default --repo https://github.com/argoproj/argocd-example-apps.git --path kustomize-guestbook --dest-namespace default --dest-server https://kubernetes.default.svc  --kustomize-replica kustomize-guestbook-ui=4 --kustomize-replica kustomize-guestbook-ui=6`
* Application
<details>
<summary>application.yaml</summary>

  ```yaml
  apiVersion: argoproj.io/v1alpha1
  kind: Application
  metadata:
    name: kustomize
    namespace: argocd
  spec:
    destination:
      namespace: default
      server: <https://kubernetes.default.svc>
    project: default
    source:
      path: kustomize-guestbook
      repoURL: <https://github.com/argoproj/argocd-example-apps.git>
      kustomize:
        replicas:
        - name: kustomize-guestbook-ui
          count: 4
  ```
</details>

* ApplicationSets
<details>
<summary>applicationset.yaml</summary>

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: guestbook
  namespace: argocd
spec:
  generators:
  - list:
      elements:
      - namespace: guestbook-qa
        url: https://kubernetes.default.svc
        values:
          replicas: "1"
      - namespace: guestbook-e2e
        url: https://kubernetes.default.svc
        values:
          replicas: "2"
      - namespace: guestbook-stage
        url: https://kubernetes.default.svc
        values:
          replicas: "3"
      - namespace: guestbook-prod
        url: https://kubernetes.default.svc
        values:
          replicas: "5"
  template:
    metadata:
      name: '{{namespace}}'
    spec:
      project: default
      source:
        repoURL: https://github.com/argoproj/argocd-example-apps.git
        targetRevision: HEAD
        path: kustomize-guestbook
        kustomize:
          replicas:
            - name: kustomize-guestbook-ui
              count: '{{values.replicas}}'
      destination:
        server: '{{url}}'
        namespace: '{{namespace}}'
      syncPolicy:
        syncOptions:
        - CreateNamespace=true
        automated:
          prune: true
          selfHeal: true
```

</details>

Signed-off-by: toyamagu2021@gmail.com <toyamagu2021@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Review by blakepettersson — APPROVED** [review]
This generally LGTM, would need to get a few more eyes on this though.

**Comment by crenshaw-dev** [comment]
@toyamagu-2021 thanks for your patience on the review! Would you mind rebasing? Looks like it needs another codegen.

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -432,6 +432,33 @@ type ApplicationSourceKustomize struct {
...
 	ForceCommonAnnotations bool `json:"forceCommonAnnotations,omitempty" protobuf:"bytes,8,opt,name=forceCommonAnnotations"`
+	// Replicas is a list of Kustomize Replicas override specifications
+	Replicas KustomizeReplicas `json:"replicas,omitempty" protobuf:"bytes,9,opt,name=replicas"`
+}
+
+type KustomizeReplica struct {
+	// Name of Deployment or StatefulSet
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+	// Number of replicas
+	Count int64 `json:"count" protobuf:"bytes,2,name=count"`
```
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/types.go:443]: I wonder if we could make this an `IntOrString`. That would make it templateable in ApplicationSets.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: pkg/apis/application/v1alpha1/types.go (hunk 1 above)
Comment (toyamagu-2021) [reply pkg/apis/application/v1alpha1/types.go]: Thank you for your review!
* I changed to IntOrString and added some tests.
* Updated PR description to include a manual test for ApplicationSets.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**